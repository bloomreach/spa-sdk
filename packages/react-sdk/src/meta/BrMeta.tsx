/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { MetaComment, Meta, META_POSITION_BEGIN, META_POSITION_END, isMetaComment } from '@bloomreach/spa-sdk';

interface BrMetaProps {
  meta: Meta[];
}

export class BrMeta extends React.Component<BrMetaProps> {
  private comments: Comment[] = [];
  private head?: Element | Text;
  private tailRef = React.createRef<HTMLElement>();

  componentDidMount() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.renderMeta();
  }

  componentDidUpdate() {
    this.head = ReactDOM.findDOMNode(this)!;
    this.removeMeta();
    this.renderMeta();
  }

  componentWillUnmount() {
    this.removeMeta();
  }

  private removeMeta() {
    this.comments.splice(0).forEach(comment => comment.remove());
  }

  private renderMeta() {
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_BEGIN)
      .forEach(this.renderMetaComment.bind(this));
    this.props.meta.filter(meta => isMetaComment(meta) && meta.getPosition() === META_POSITION_END)
      .reverse()
      .forEach(this.renderMetaComment.bind(this));
  }

  private renderMetaComment(meta: MetaComment) {
    const comment = this.head!.ownerDocument!.createComment(meta.getData());
    this.comments.push(comment);

    if (meta.getPosition() === META_POSITION_BEGIN) {
      return void this.head!.parentNode!.insertBefore(comment, this.head!);
    }

    if (!this.tailRef.current!.nextSibling) {
      return void this.tailRef.current!.parentNode!.appendChild(comment);
    }

    this.tailRef.current!.parentNode!.insertBefore(comment, this.tailRef.current!.nextSibling);
  }

  render() {
    return (
      <>
        {this.props.children}
        {this.props.meta.length ? <span style={{ display: 'none' }} ref={this.tailRef} /> : null}
      </>
    );
  }
}
