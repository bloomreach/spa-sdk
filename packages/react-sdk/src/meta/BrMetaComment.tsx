/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { MetaComment } from '@bloomreach/spa-sdk';

interface BrMetaCommentProps {
  meta: MetaComment;
}

export class BrMetaComment extends React.Component<BrMetaCommentProps> {
  private placeholder = React.createRef<HTMLElement>();
  private comment?: Comment;

  componentDidMount() {
    const placeholder = this.placeholder.current!;
    this.comment = placeholder.ownerDocument!.createComment(this.props.meta.getData());
    placeholder.parentNode!.replaceChild(this.comment, placeholder);
  }

  shouldComponentUpdate(nextProps: BrMetaCommentProps) {
    return nextProps.meta.getData() !== this.props.meta.getData();
  }

  componentDidUpdate() {
    this.comment!.replaceData(0, this.comment!.data.length, this.props.meta.getData());
  }

  componentWillUnmount() {
    this.comment!.parentNode!.replaceChild(this.placeholder.current!, this.comment!);
    delete this.comment;
  }

  render() {
    return <span style={{ display: 'none' }} ref={this.placeholder} />;
  }
}
