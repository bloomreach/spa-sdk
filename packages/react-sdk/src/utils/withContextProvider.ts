// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';
import React, { ReactElement } from 'react';

export const withContextProvider = (context?: Record<string, any>, children?: ReactElement): React.ReactElement => {
  const childContextTypes = (): Record<string, string> | undefined => context
    && Object.keys(context).reduce((obj: Record<string, any>, key: string) => {
      obj[key] = PropTypes.any;
      return obj;
    }, {});

  class ContextProvider extends React.Component {
    getChildContext(): Record<string, string> | undefined {
      return context;
    }

    render(): React.ReactNode {
      return children;
    }
  }

  (ContextProvider as any).displayName = 'ContextProvider';
  (ContextProvider as any).childContextTypes = childContextTypes();

  return React.createElement(ContextProvider, null);
};
