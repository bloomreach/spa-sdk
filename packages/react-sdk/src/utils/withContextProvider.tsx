/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { ReactElement } from 'react';
import { BrPageContext } from '../page';
import { BrComponentContext, BrMappingContext } from '../component';

export const withPageContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrPageContext.Provider value={context}>{children}</BrPageContext.Provider>;
};

export const withMappingContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return (
    <BrMappingContext.Provider value={context}>
      {children}
    </BrMappingContext.Provider>
  );
};

export const withComponentContextProvider = (context?: any, children?: ReactElement): React.ReactElement => {
  if (!context) {
    return children as React.ReactElement;
  }

  return <BrComponentContext.Provider value={context}>{children}</BrComponentContext.Provider>;
};
