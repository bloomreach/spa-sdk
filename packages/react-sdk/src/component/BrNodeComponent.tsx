/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Component } from '@bloomreach/spa-sdk';
import React, { useContext } from 'react';
import { BrMeta } from '../meta';
import { BrMappingContext } from './BrMappingContext';
import { BrProps } from './BrProps';

export function BrNodeComponent<T extends Component>(props: React.PropsWithChildren<BrProps<T>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const mapping = component && (context[component.getName()] as React.ComponentType<BrProps>);
  const meta = component?.getMeta();

  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
