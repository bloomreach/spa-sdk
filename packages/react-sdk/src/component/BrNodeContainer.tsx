/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useContext } from 'react';
import {
  Container,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
} from '@bloomreach/spa-sdk';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';
import {
  BrContainerBox,
  BrContainerInline,
  BrContainerNoMarkup,
  BrContainerOrderedList,
  BrContainerUnorderedList,
} from '../cms';

export function BrNodeContainer(props: React.PropsWithChildren<BrProps<Container>>): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, children } = props;

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    switch (type) {
      case TYPE_CONTAINER_INLINE:
        return BrContainerInline;
      case TYPE_CONTAINER_NO_MARKUP:
        return BrContainerNoMarkup;
      case TYPE_CONTAINER_ORDERED_LIST:
        return BrContainerOrderedList;
      case TYPE_CONTAINER_UNORDERED_LIST:
        return BrContainerUnorderedList;
      default:
        return BrContainerBox;
    }
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
