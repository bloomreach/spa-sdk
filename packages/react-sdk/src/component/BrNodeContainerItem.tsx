/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useContext, useEffect, useState } from 'react';
import { ContainerItem, TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { BrContainerItemUndefined } from '../cms';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';

export function BrNodeContainerItem(
  props: React.PropsWithChildren<BrProps<ContainerItem>>,
): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, page, children } = props;
  const [forcedRerenders, setForcedRerenders] = useState(0);

  const onUpdate = useCallback((): void => {
    // Trigger a rerender of this component due to a non-react 3rd party event
    setForcedRerenders((c) => c + 1);
  }, []);

  useEffect(() => {
    component?.on('update', onUpdate);

    return () => {
      component?.off('update', onUpdate);
    };
  }, [component, onUpdate]);

  useEffect(() => {
    // Skip the mount event page sync as BrPage just called it
    if (forcedRerenders > 0) {
      page?.sync();
    }
  }, [page, forcedRerenders]);

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    return (
      (context[
        TYPE_CONTAINER_ITEM_UNDEFINED as any
      ] as React.ComponentType<BrProps>) ?? BrContainerItemUndefined
    );
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
