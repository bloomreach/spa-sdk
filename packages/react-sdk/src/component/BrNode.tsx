/*
 * Copyright 2019-2020 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isContainer, isContainerItem, Component } from '@bloomreach/spa-sdk';
import { BrMeta } from '../meta';
import { BrPageContext } from '../page/BrPageContext';
import { BrComponentContext } from './BrComponentContext';
import { BrNodeContainer } from './BrNodeContainer';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { BrNodeComponent } from './BrNodeComponent';

interface BrNodeProps {
  component?: Component;
}

export class BrNode extends React.Component<BrNodeProps> {
  static contextType = BrPageContext;

  context: React.ContextType<typeof BrPageContext>;

  private renderNode(): JSX.Element | React.ReactNode {
    const { children, component } = this.props;

    if (React.Children.count(children)) {
      return <BrMeta meta={component?.getMeta()}>{children}</BrMeta>;
    }

    // eslint-disable-next-line react/no-array-index-key
    const childrenList = component?.getChildren().map((child, index) => <BrNode key={index} component={child} />);

    if (isContainer(component)) {
      return (
        <BrNodeContainer component={component} page={this.context!}>
          {childrenList}
        </BrNodeContainer>
      );
    }

    if (isContainerItem(component)) {
      return (
        <BrNodeContainerItem component={component} page={this.context!}>
          {childrenList}
        </BrNodeContainerItem>
      );
    }

    return (
      <BrNodeComponent component={component} page={this.context!}>
        {childrenList}
      </BrNodeComponent>
    );
  }

  render(): JSX.Element {
    const { component } = this.props;

    return <BrComponentContext.Provider value={component}>{this.renderNode()}</BrComponentContext.Provider>;
  }
}
