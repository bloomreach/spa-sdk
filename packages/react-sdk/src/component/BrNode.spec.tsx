/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { isContainer, isContainerItem, Component, MetaCollection, Page } from '@bloomreach/spa-sdk';
import { render } from '@testing-library/react';
import { BrNode } from './BrNode';
import { BrNodeComponent } from './BrNodeComponent';
import { BrNodeContainer } from './BrNodeContainer';
import { BrNodeContainerItem } from './BrNodeContainerItem';
import { withContextProvider } from '../utils/withContextProvider';
import { BrMeta } from '../meta';

jest.mock('@bloomreach/spa-sdk');

describe('BrNode', () => {
  const context = {
    isPreview: jest.fn(),
  } as unknown as jest.Mocked<Page>;
  const props = {
    component: {
      getChildren: jest.fn(() => []),
      getMeta: jest.fn(() => []),
      getName: jest.fn(),
      getType: jest.fn(),
      on: jest.fn(),
      off: jest.fn(),
    } as unknown as jest.Mocked<Component>,
  };

  beforeEach(() => {
    jest.restoreAllMocks();

    (BrNode as any).contextTypes = {
      isPreview: () => null,
    };
    delete (BrNode as Partial<typeof BrNode>).contextType;
  });

  it('should render a component meta-data', () => {
    const meta = {} as MetaCollection;
    props.component.getMeta.mockReturnValueOnce(meta);
    const element = render(<BrNode {...props} />);

    const nodeMeta = render(withContextProvider(context, <BrMeta meta={props.component.getMeta()} />));

    expect(element.container.isEqualNode(nodeMeta.container)).toBe(true);
    expect(element.asFragment()).toMatchSnapshot();
  });

  it('should render a component', () => {
    const element = render(withContextProvider(context, <BrNode {...props} />));

    const nodeComponent = render(withContextProvider(context, <BrNodeComponent {...props} />));

    expect(element.container.isEqualNode(nodeComponent.container)).toBe(true);
    expect(element.asFragment()).toMatchSnapshot();
  });

  it('should render a container', () => {
    jest.mocked(isContainer).mockReturnValueOnce(true);
    const element = render(withContextProvider(context, <BrNode {...props} />));

    const nodeContainer = render(<BrNodeContainer page={context} component={props.component as any} />);

    expect(element.container.isEqualNode(nodeContainer.container)).toBe(true);
    expect(element.asFragment()).toMatchSnapshot();
  });

  it('should render a container item', () => {
    jest.mocked(isContainerItem).mockReturnValueOnce(true);
    const element = render(withContextProvider(context, <BrNode {...props} />));

    const nodeContainerItem = render(<BrNodeContainerItem page={context} component={props.component as any} />);

    expect(element.container.isEqualNode(nodeContainerItem.container)).toBe(true);
    expect(element.asFragment()).toMatchSnapshot();
  });

  it('should render children if present', () => {
    const element = render(
      <BrNode {...props}>
        <a />
        <b />
      </BrNode>,
    );

    expect(element.asFragment()).toMatchSnapshot();
  });

  it('should render model children if component children are not present', () => {
    const component1 = { ...props.component } as Component;
    const component2 = { ...props.component } as Component;
    props.component.getChildren.mockReturnValueOnce([component1, component2]);
    const element = render(<BrNode {...props} />);

    expect(element.asFragment()).toMatchSnapshot();
  });
});
