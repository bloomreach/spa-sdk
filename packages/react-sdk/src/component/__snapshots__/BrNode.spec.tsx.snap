// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrNode should render a component 1`] = `<DocumentFragment />`;

exports[`BrNode should render a component meta-data 1`] = `<DocumentFragment />`;

exports[`BrNode should render a container 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`BrNode should render a container item 1`] = `
<DocumentFragment>
  <div>
    Component "undefined" is not defined.
  </div>
</DocumentFragment>
`;

exports[`BrNode should render children if present 1`] = `
<DocumentFragment>
  <a />
  <b />
</DocumentFragment>
`;

exports[`BrNode should render model children if component children are not present 1`] = `<DocumentFragment />`;
